### DISCUSSION TITLE
ESP32 TCP Client and multiple Server

### DISCUSSION THREAD

**Discussion by diegostf** [eModbus/eModbus#270] [Q&A] [answered]
I have one ESP32 (TCP Client) and ten sensors (ESP8266 TCP server). What better way to get/read the data?
Can I send requests to everyone (one by one) and handle the information as it arrives? Or do I need to make a request and wait for the data before making the next one?

**Comment by Miq1** [comment, marked as answer]
It is quite normal to have several servers of course. The library is built to serialize the requests internally. You can fire your ``addRequest()`` calls as fast as you are able on the client. The requests are put on a queue and sent in the order you added them. The library will take care of the queue, and as soon as a response has been received (or if the timeout has struck) it will be forwarded to your ``onData`` or ``onError`` handlers. If some server will answer earlier than another, the sequence of responses may be different from the sequence of your requests. You only can avoid that by using the ``syncRequest()`` calls instead, should you require it.

**Reply by diegostf** [reply]
Do both libraries do this (sync and async)? I read in doc: "The async version will connect to exactly one Modbus server."
The answers, even out of order, will be treated by the token.

**Reply by Miq1** [reply]
That again is subject to change in the Async TCP version. The synchronous (aka ``ModbusClientTCP``) has a ``setTarget()`` method to address another server for the forthcoming requests that will be adapted to the Async TCP version as well.

As a workaround you may use the AsyncTCP ``connect()`` method to switch servers.

**Reply by diegostf** [reply]
What is the difference between the libraries (sync and async)? Reading the documentation, I got confused.

**Reply by Miq1** [reply]
We must not mangle sync/async **requests** and sync/async **TCP** here.

### Requests
A Modbus request sent with ``addRequest()`` always is asynchronous. It will be put on a queue, sent and processed in sequential manner and as soon as e response is received or the timeout has passed without any answer, the ``onResponse``, ``onData`` or ``onError`` handlers are called (that you provided).
Opposite to that, a request sent by ``syncRequest()`` will be sent and while it is processed your code will be held. Only if a response is received (or timeout has struck) you will get control again. This is called a synchronous request.

### TCP
The normal case for the ESP32 is the synchronous TCP handling. It boasts a background task that is operating on a standard TCP stack. It freely can operate on the queue and serve the requests and responses in order as are coming in.
The ESP8266 lacks a proper multitasking, so the synchronous TCP would not work without blocking the user code while a request is being processed. To get around that, the Async TCP library was created by me-no-dev, that is in widespread use and is doing a specialized light multitasking just for the TCP communications. The eModbus library is using that for the asyncTCP parts of the library to enable the application on the ESP8266 as well. Here the way the TCP may be used is much more restrictive than on a full-size ESP32. Thus checks of the request queue or if responses have been received can be done only in intervals (500ms in our case), plus some TCP features simply do not exist - like immediate detection of an irresponsive server.
Concluding on an ESP8266 AsyncTCP is the only way to go, but for the cost of far less throughput or responsiveness.

**Reply by diegostf** [reply]
Tks Niq1. Now it's running using ModbusClientTCP and addRequest() calls, requesting data from 5 ESP8266 servers. The response time is between 40 to 700ms.
